File: test/language/templates/cpp/ChoiceTemplatedEnumSelectorTest.cpp
#include "gtest/gtest.h"

#include "templates/choice_templated_enum_selector/ChoiceTemplatedEnumSelector.h"

#include "zserio/RebindAlloc.h"

namespace templates
{
namespace choice_templated_enum_selector
{

using allocator_type = ChoiceTemplatedEnumSelector::allocator_type;
using string_type = zserio::string<allocator_type>;

TEST(ChoiceTemplatedEnumSelectorTest, readWrite)
{
    ChoiceTemplatedEnumSelector choiceTemplatedEnumSelector;
    choiceTemplatedEnumSelector.setSelectorFromZero(EnumFromZero::ONE);
    choiceTemplatedEnumSelector.setSelectorFromOne(EnumFromOne::THREE);
    {
        TemplatedChoice_EnumFromZero fromZeroChoice;
        fromZeroChoice.setUint16Field(42);
        choiceTemplatedEnumSelector.setFromZeroChoice(std::move(fromZeroChoice));
    }
    {
        TemplatedChoice_EnumFromOne fromOneChoice;
        fromOneChoice.setStringField("string");
        choiceTemplatedEnumSelector.setFromOneChoice(fromOneChoice); // copy
    }
    choiceTemplatedEnumSelector.initializeChildren();

    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
    zserio::BitStreamWriter writer(bitBuffer);
    choiceTemplatedEnumSelector.write(writer);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    ChoiceTemplatedEnumSelector readChoiceTemplatedEnumSelector(reader);

    ASSERT_TRUE(choiceTemplatedEnumSelector == readChoiceTemplatedEnumSelector);
}

} // namespace choice_templated_enum_selector
} // namespace templates

File: test/language/templates/python/ChoiceTemplatedEnumSelectorTest.py
import zserio

import Templates

class ChoiceTemplatedEnumSelectorTest(Templates.TestCase):
    def testReadWrite(self):
        selectorFromZero = self.api.EnumFromZero.ONE
        selectorFromOne = self.api.EnumFromOne.THREE
        fromZeroChoice = self.api.TemplatedChoice_EnumFromZero(selectorFromZero)
        fromZeroChoice.uint16_field = 42
        fromOneChoice = self.api.TemplatedChoice_EnumFromOne(selectorFromOne, string_field_="string")
        choiceTemplatedEnumSelector = self.api.ChoiceTemplatedEnumSelector(
            selectorFromZero, selectorFromOne, fromZeroChoice, fromOneChoice
        )

        writer = zserio.BitStreamWriter()
        choiceTemplatedEnumSelector.write(writer)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readChoiceTemplatedEnumSelector = self.api.ChoiceTemplatedEnumSelector()
        readChoiceTemplatedEnumSelector.read(reader)
        self.assertEqual(choiceTemplatedEnumSelector, readChoiceTemplatedEnumSelector)
